src/check/arbitrary/LetRecArbitrary.ts

```diff
@@ -74,7 +79,7 @@ function isLazyArbitrary(arb: Arbitrary<any> | undefined): arb is LazyArbitrary
  * @public
  */
 export function letrec<T>(
-  builder: (tie: (key: string) => Arbitrary<unknown>) => { [K in keyof T]: Arbitrary<T[K]> }
+  builder: (tie: TieFunction<T>) => { [K in keyof T]: Arbitrary<T[K]> }
```

Comment:
From what I see from the tests only targeting typings (see https://github.com/dubzzz/fast-check/blob/main/test/type/main.ts), the change does not seem to be backward compatible in its current shape.

We have errors like:
```txt
Error: main.ts(291,3): error TS2345: Argument of type '"No recursion \"letrec\""' is not assignable to parameter of type '{ expected: { a: Arbitrary<number>; b: Arbitrary<string>; }; got: {}; }'.
Error: main.ts(298,3): error TS2345: Argument of type '"Recursive \"letrec\""' is not assignable to parameter of type '{ expected: { a: Arbitrary<number>; b: Arbitrary<unknown>; }; got: {}; }'.
Error: main.ts(305,3): error TS2345: Argument of type '"Invalid recursion \"letrec\""' is not assignable to parameter of type '{ expected: { a: Arbitrary<number>; b: Arbitrary<unknown>; }; got: {}; }'.
```

Reply:
<del>This is very unfortunate. I've tried investigating and debugging it, but I have ran out of time. As far as I can see, as soon as generic `T` is used somewhere, it starts participating in the TypeScript inference. Maybe, if it was possible to mark an argument as excluded from participating in inference, it would work (but this is just a guess). There is a pretty old issue about that https://github.com/microsoft/TypeScript/issues/14829 too. However, when I tried some of the ideas in that issue, they do not work. There is weirdness that if you don't use `tie` function in the callback, then everything works, but as soon as parameter to callback is specified, it breaks.</del>

EDIT:
As a last ditch effort, I tried to put Cb into generic, and it seems to work? I hope... :) 
The issue that might be is that someone could explicitly overwrite that Cb generic, but that might be noted as an unsupported use-case.

Reply:
Okay, so now only 3.2 fails in CI, I'm not sure why, it seems that locally 3.2 passes for me?

Reply:
@dubzzz Do you have any ideas why only 3.2 might be failing? Thanks.

Reply:
No idea for the moment, I'll try to investigate by checkouting the branch locally 🤔 (really sorry, I missed your previous message)